Feature/upgrade project by comments

TazmanianDevil opened the pull request.

TazmanianDevil committed:
Исправление части замесаний
1. Убран интерфейс CounterService;
2. Поправлен метод CounterService.getNames();
3. Убран ненужный Spring Config

TazmanianDevil committed:
Try to make more RESTful API

taiberium commented on src/main/java/ru/tazik/counters/dto/CounterUpdateRq.java:
```diff
@@ -0,0 +1,9 @@
+import ru.tazik.counters.enums.CounterCommand;
+
+@Data
+public class CounterUpdateRq {
```
Rq это Request? :-) Стоит ли экономить буквы?

TazmanianDevil replied:
Переименовал :)

taiberium commented on src/main/java/ru/tazik/counters/dto/CounterUpdateRq.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Data;
+import ru.tazik.counters.enums.CounterCommand;
+
+@Data
```
У  Lombok есть аннотация @Value, которая фактически делает объект иммутабельным, стоит использовать @Value везде где это возможно и стремиться к иммутабельности.

taiberium commented on src/main/java/ru/tazik/counters/endpoint/CounterController.java:
```diff
@@ -1,53 +1,67 @@
 
 @RestController
 @RequestMapping("counters")
+@AllArgsConstructor
```
Стоит использровать @RequiredArgsConstructor, эта аннотация создаст конструтктор только для final полей, это предотвратит проблемы если у тебя есть еще и field injection c не final полями по каким-то причинам.

TazmanianDevil replied:
Убрал. Но так как Jackson для десериализации требует дефолтный конструктор, пришлось сделать билдер.

taiberium commented on src/main/java/ru/tazik/counters/endpoint/CounterController.java:
```diff
@@ -1,53 +1,67 @@
-    @Autowired
-    public CounterController(CounterService counterService) {
-        this.counterService = counterService;
+    @RequestMapping(method = RequestMethod.POST, consumes = MediaType.APPLICATION_JSON,
```
Используй @PostMapping consumes и produces по умолчанию и так будет Json, не обязательно указывать :-)
Можно было написать просто @PostMapping

TazmanianDevil replied:
Спасибо, исправил

taiberium commented on src/main/java/ru/tazik/counters/endpoint/CounterController.java:
```diff
@@ -1,53 +1,67 @@
+        String name = counterService.create();
+        URI uri = ServletUriComponentsBuilder.fromCurrentServletMapping().path("/counters/{id}").build()
+                .expand(name).toUri();
+        return ResponseEntity.created(uri).build();
```
Этот момент хотелось бы обсудить вживую, не совсем понятно зачем так поступил))

TazmanianDevil replied:
https://restfulapi.net/http-status-201-created/

taiberium commented on src/main/java/ru/tazik/counters/service/CounterService.java:
```diff
@@ -1,46 +1,89 @@
+            counters.put(name, counter);
+            return name;
+        } else {
+            return create();
```
Рекурсия это плохо, при самом неудачном сценарии, Stackoverflow Exception :-)

TazmanianDevil replied:
Исправил

TazmanianDevil committed:
Исправление замечаний v.2
1. Поправлены маппинги рестов
2. Доведен до ума обработчик исключений
3. Убран ResponseEntity там, где он не нужен
